### PR TITLE
Implement TextDocument/rangeFormatting

### PR DESCRIPTION
**Pull request by mkcms**
This commit implements the TextDocument/rangeFormatting method.

I have already signed and sent the FSF copyright assignment papers, but unfortunately the copyright clerk left FSF soon after that, so it will probably take a while until the whole process is finished.

### CODE REVIEW

File: eglot.el [hunk 1]
```diff
@@ -976,6 +976,11 @@ THINGS are either registrations or unregisterations."
   (list :textDocument (eglot--TextDocumentIdentifier)
         :position (eglot--pos-to-lsp-position)))
 
+(defun eglot--FormattingOptions ()
```
Comment (joaotavora) [inline eglot.el]: Shouldn't this be reused in `eglot-format-buffer`?

File: eglot.el [hunk 2]
```diff
@@ -1194,11 +1199,24 @@ DUMMY is ignored."
     (eglot--current-server-or-lose)
     :textDocument/formatting
     (list :textDocument (eglot--TextDocumentIdentifier)
-          :options (list :tabSize tab-width
-                         :insertSpaces
-                         (if indent-tabs-mode :json-false t)))
+          :options (eglot--FormattingOptions))
     :deferred :textDocument/formatting)))
 
+(defun eglot-format (beg end)
```
Comment (joaotavora) [inline eglot.el]: This command is very very similar to `eglot-format-buffer`. I propose we get rid of `eglot-format-buffer`, and if the is no region selected, default the BEG and END to in `eglot-format`, and proceed according to the old implementation of `eglot-format-buffer`.  Then make `eglot-format-buffer` an obsolete alias of `eglot-format`.

Comment (mkcms) [reply eglot.el]: So should support `:textDocument/formatting` be dropped entirely in favor of `:textDocument/rangeFormatting`?

File: eglot.el (hunk 1 above)
Comment (mkcms) [reply eglot.el]: Isn't it?

File: eglot.el (hunk 2 above)
Comment (joaotavora) [reply eglot.el]: No, no.  Just choose the former when there is no selected region, and the latter when there is.  Just a suggestion, don't you like it?

File: eglot.el (hunk 1 above)
Comment (joaotavora) [reply eglot.el]: Oh right :flushed:, sorry

**Comment by joaotavora** [comment]
> I chose not to make eglot-format-buffer an obsolete alias of eglot-format, because actually it's behavior is a little bit different - currently it always formats the entire buffer even if region is active.

Hmmm, OK (though we could afford the backward incompatibility probably).

But I also want `eglot-format` to use the "whole file" behaviour if there is no active regino.  Are you sure the `(interactive "r")` spec does that?  Shouldn't we use some

```
(interactive (and (region-active-p)
                        (list (region-beginning) (region-end))))
```

instead?  (and make the BEG and END `&optional`).

**Comment by mkcms** [comment]
Ok, I changed `eglot-format` so that it formats entire buffer if called interactively without active region, like you suggested.

**Comment by mkcms** [comment]
> I think we might still be missing:
> 
> a capability entry in eglot-client-capabilities (not sure)
> a tick in the checkbox in README.md
> a mention of eglot-format in README.md (you can replace the eglot-format-buffer entry)

Ok.

**Comment by mkcms** [comment]
> there's a bug in either python-lsp or in eglot which I will report later.

Turns out it was a bug in Emacs which I reported and is already fixed: [32237](https://debbugs.gnu.org/cgi/bugreport.cgi?bug=32237). The tests I think should be merged into one, but I don't know if the CI server will have Emacs version which contains a fix for this bug?

**Comment by mkcms** [comment]
> Shouldn't we have a workaround in place for 26.1?

I think so. Perhaps there should be some `eglot--inhibit-change-hooks` variable set from `eglot--apply-text-edits` which will make `eglot--before-change` ignore changes so that `eglot--apply-text-edits` can record them manually.

I just compiled emacs-26 branch locally and tested eglot on it. It turns out that the bug may not be fixed after all. The arguments passed to `before-change-functions` don't seem to cover the entire changed region of the buffer.

**Comment by mkcms** [comment]
> Perhaps there should be some eglot--inhibit-change-hooks

I see now that `eglot--recent-changes` can just be set to `t` temporarily, I think.

**Comment by joaotavora** [comment]
> eglot--apply-text-edits can record them manually.

Hmmm, this would be the last resort I think.  A better alternative would be to *not* use `replace-buffer-contents` until it is working correctly (and keep reporting the bugs until it is).  We would, of course, lose the advantage that we selected it for, which is keeping point reasonably stable, but I think that's a lower price to pay than a malfunctioning recording of changes, or a hacked up eglot.el.

**Comment by mkcms** [comment]
> A better alternative would be to not use replace-buffer-contents until it is working correctly

It's your choice, but personally I like the fact that this function preserves more than just point position, i.e. bookmark overlays or folds as I use them all the time.

> I think that's a lower price to pay than a malfunctioning recording of changes, or a hacked up eglot.el.

See end of this comment for a simple hack that seems to fix this issue. IMO it's not that bad. But again, it's your choice.

---

I tested `replace-buffer-contents` some more and I think that the problem now (when Emacs bug [32237](https://debbugs.gnu.org/cgi/bugreport.cgi?bug=32237) is fixed) is that this function doesn't call `before-change-functions` at all in some situations, but it calls `after-change-functions` . And when it does call both functions, there is difference between the second parameter in both calls (I only tested some specific scenarios).

When `replace-buffer-contents` doesn't call `before-change-functions`, eglot signals an error:
```
Debugger entered--Lisp error: (wrong-type-argument consp nil)
  setcdr(nil (0 "  pass\n"))
  (let* ((#:v (car eglot--recent-changes))) (setcdr (cdr #:v) (list pre-change-length (buffer-substring-no-properties start end))))
  (if (and (listp eglot--recent-changes) (null (cdr (cdr (car eglot--recent-changes))))) (let* ((#:v (car eglot--recent-changes))) (setcdr (cdr #:v) (list pre-change-length (buffer-substring-no-properties start end)))) (setq eglot--recent-changes :emacs-messup))
  eglot--after-change(31 38 0)
  replace-buffer-contents(#<buffer  *temp*>)
  (save-restriction (narrow-to-region beg end) (replace-buffer-contents temp))
  (save-excursion (save-restriction (narrow-to-region beg end) (replace-buffer-contents temp)))
```

Recipe for this backtrace (on eglot master):

1. Create a python file with these contents:

    ```python
    def foo():pass
    def bar():pass
    ```

2. `C-x C-f` this file (kill buffer first so that `eglot--recent-changes` is nil)
3. `M-x eglot`
4. `M-x eglot-format-buffer`

After the error, it will probably not happen again, because `eglot--after-change` is removed from `after-change-functions`.

A quick hack that seems to fix these issues is below.  I haven't tested it much.

```diff
diff --git a/eglot.el b/eglot.el
index 1d8f2eb..a0d05cd 100644
--- a/eglot.el
+++ b/eglot.el
@@ -1417,7 +1417,10 @@ If SKIP-SIGNATURE, don't try to send textDocument/signatureHelp."
                       (save-excursion
                         (save-restriction
                           (narrow-to-region beg end)
-                          (replace-buffer-contents temp)))
+                          (let ((inhibit-modification-hooks t))
+                            (eglot--before-change beg end)
+                            (replace-buffer-contents temp)
+                            (eglot--after-change beg (+ beg (length newText)) (- end beg)))))
                       (progress-reporter-update reporter (cl-incf done)))))))
             (mapcar (jsonrpc-lambda (&key range newText)
                       (cons newText (eglot--range-region range 'markers)))

```

**Comment by joaotavora** [comment]
> It's your choice, but personally I like the fact that this function preserves more than just point position, i.e. bookmark overlays or folds as I use them all the time.

Indeed, that's useful.  But if it's broken... I gotta see your workaround and understand the bug/risks better, and perhaps it should stay.

**Comment by joaotavora** [comment]
> A quick hack that seems to fix these issues is below. I haven't tested it much.

Indeed, it look safe enough.  Can you run it by emacs-devel to get Eli and Stefan's opinion?

**Comment by mkcms** [comment]
> Can you run it by emacs-devel to get Eli and Stefan's opinion?

Sure, I will report the bug and ask a question separately on emacs-devel later.

File: README.md [hunk 3]
```diff
@@ -96,8 +96,8 @@ Here's a summary of available commands:
 
 - `M-x eglot-rename` ask the server to rename the symbol at point;
 
-- `M-x eglot-format-buffer` ask the server to reformat the current
-  buffer;
+- `M-x eglot-format` asks the server to format active region, or
```
Comment (joaotavora) [inline README.md]: "_the_ active region".  But isn't it simpler to say "asks the server to format the buffer or the active region".

File: eglot.el [hunk 4]
```diff
@@ -1227,17 +1228,35 @@ DUMMY is ignored."
...
+                                  :end (eglot--pos-to-lsp-position end)))))
+                (t
+                 '(:textDocument/formatting :documentFormattingProvider nil)))))
+    (unless (eglot--server-capable cap)
+      (eglot--error "Server can't format!"))
+    (eglot--apply-text-edits
+     (jsonrpc-request
+      (eglot--current-server-or-lose)
+      method
+      (apply #'list
```
Comment (joaotavora) [inline eglot.el]: `apply #'list` is very akward.  Use a backquote expression or `cl-list*`.

Comment (joaotavora) [reply eglot.el]: in case you need a reminder
```
`(:thing1 ,thing1 :thing2 ,thing2 ,@args)
```

**Comment by mkcms** [comment]
OK, I pushed fixes for those issues.

**Comment by joaotavora** [comment]
I merged this manually by cherry-picking the second commit onto master.  This is because I had already pulled in the other bugfix commit and I didn't want any duplicates.

**Pull request closed without merging**